pkg/replica/hash.go

```diff
@@ -0,0 +1,340 @@
+	"syscall"
+	"time"
+
+	"github.com/gofrs/flock"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+
+	diskutil "github.com/longhorn/longhorn-engine/pkg/util/disk"
+)
+
+const (
+	defaultHashMethod = "crc64"
+
+	xattrSnapshotHashName     = "user.longhorn.metadata"
+	xattrSnapshotHashValueMax = 256
+
+	fileLockDirectory = "/host/var/lib/longhorn/.lock"
+	hashLockFileName  = "hash"
+)
+
+type SnapshotHashStatus struct {
+	StatusLock sync.RWMutex
+
+	State             ProgressState
+	Checksum          string
+	Error             string
+	SilentlyCorrupted bool
+}
+
+type SnapshotHashTask struct {
+	sync.Mutex
+
+	Ctx        context.Context
+	CancelFunc context.CancelFunc
+
+	SnapshotName string
+	Rehash       bool
+
+	file *os.File
+
+	SnapshotHashStatus
+}
+
+type SnapshotXattrHashInfo struct {
+	Method       string `json:"method"`
+	Checksum     string `json:"checksum"`
+	ModTime      string `json:"modTime"`
+	LastHashedAt string `json:"lastHashedAt"`
+}
+
+func NewSnapshotHashTask(ctx context.Context, cancel context.CancelFunc, snapshotName string, rehash bool) *SnapshotHashTask {
+	return &SnapshotHashTask{
+		Ctx:          ctx,
+		CancelFunc:   cancel,
+		SnapshotName: snapshotName,
+		Rehash:       rehash,
+
+		SnapshotHashStatus: SnapshotHashStatus{
+			State: ProgressStateInProgress,
+		},
+	}
+}
+
+func (t *SnapshotHashTask) LockFile() (fileLock *flock.Flock, err error) {
+	defer func() {
+		if err != nil && fileLock != nil && fileLock.Path() != "" {
+			if err := os.RemoveAll(fileLock.Path()); err != nil {
+				logrus.Warnf("failed to remove lock file %v since %v", fileLock.Path(), err)
+			}
+		}
+	}()
+
+	hostName := os.Getenv("HOSTNAME")
+	if hostName == "" {
+		return nil, fmt.Errorf("env HOSTNAME is missing")
+	}
+
+	err = os.MkdirAll(fileLockDirectory, 0755)
+	if err != nil {
+		return nil, err
+	}
+
+	fileLock = flock.New(filepath.Join(fileLockDirectory, hashLockFileName))
+
+	// Blocking lock
+	err = fileLock.Lock()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to fetch the file lock for hashing snapshot %v", t.SnapshotName)
+	}
+
+	return fileLock, nil
+}
+
+func (t *SnapshotHashTask) UnlockFile(fileLock *flock.Flock) {
+	fileLock.Unlock()
+}
+
+func (t *SnapshotHashTask) Execute() {
+	var err error
+	var checksum string
+
+	defer func() {
+		t.StatusLock.Lock()
+		defer t.StatusLock.Unlock()
+		t.Checksum = checksum
+		if err != nil {
+			logrus.Errorf("failed to hash snapshot %v since %v", t.SnapshotName, err)
+			t.State = ProgressStateError
+			t.Error = err.Error()
+		} else {
+			t.State = ProgressStateComplete
+		}
+	}()
+
+	// Each node can have only snapshot hashing task at the same time.
+	// When the snapshot hashing task is initiated, the task will try to fetch the
+	// file lock (${fileLockDirectory}/hash), and other tasks need to wait for the lock.
+	// The file is unlocked after the task is completed.
+	fileLock, err := t.LockFile()
+	if err != nil {
+		return
+	}
```

Comment:
If the file locking failed (there is already one hashing in-progress, what will the caller handle it? Retry? Error out? Or stuck there?

Reply:
This is a blocking lock. If anothoer hashing in-progress, it will wait for the lock. If the error is caused by another issue such as EBADF, the hashing fails.

Reply:
> Can I say that all of the workers have a chance to send a hash request to a specific node? Then how about 100 nodes encountering the same scenario?

Yeah, in the case, some nodes might be idle. Howerever, if the volumes and replicas are evenly distributed, the idle cases should be reduced a lot.

Reply:
I mean, in the example mentioned above, there may be 500 goroutinues getting stuck in this locking...

Reply:
I see. Would you mind reducing the 5 task workers to 1?
Probably we can drop the request if the number of waiting tasks reaches to the limit. The number of waiting tasks can be recorded in the file lock's content.

Reply:
I think it's better to return an error here rather than keep waiting. Then the longhorn manager will requeue the task with a delay. But I am not sure if it would be too complicated.

Reply:
So, no waiting tasks and requeue the tasks that cannot get the lock. Correct?

Reply:
As long as there is no tons of task waiting there, I am OK with any other solutions.

Reply:
Addressed. Thank you.

Reply:
Can you point out the related modifications? BTW, how do you requeue the task if there is an existing in-progress task? I didn't find the code in the longhorn-manager PR.

Reply:
Before requesting, engine will check lock state of the replica's node ([codes](https://github.com/longhorn/longhorn-engine/blob/a5df8bb33eb46416120b20f525140abf2159e84d/pkg/sync/sync.go#L909)). If the lock file is held by another task, then engine returns the error `cannot request hashing snapshot because %s is hashing snapshot`.

In snashot monitor side, if the error `cannot request hashing snapshot` is received, then requeue the failed request ([codes](https://github.com/longhorn/longhorn-manager/blob/717cbb177393dca64d409e5c7d9d38cd085be7ec/controller/monitor/snapshot_monitor.go#L215))

Reply:
OK. The target sync agent server will check if other servers (of other replicas) are hashing... 
As we discussed before, can we move this to the instance manager part? cc @joshimoo

Reply:
Engine directly sends the requests to replicas' sync-agent servers, so handling the logic in IM-r is not available. If we implement the replica proxy in the future (i.e engine process->replica proxy->replica process), we can move it to IM-r and record the count in the IM-r. Let's improve it in the future. WDYT?